Deprecated `std::iterator` use

On new enough toolchains:

```plain
/spread/mir/src/platforms/common/server/kms-utils/drm_mode_resources.h:61:21: error: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Werror=deprecated-declarations]
   61 |         public std::iterator<std::input_iterator_tag, DRMUPtr>
      |                     ^~~~~~~~
In file included from /usr/include/c++/12/bits/stl_algobase.h:65,
                 from /usr/include/c++/12/memory:63,
                 from /spread/mir/src/platforms/common/server/kms-utils/drm_mode_resources.h:25:
/usr/include/c++/12/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
```

This is ignored in #2296, but the proper fix would be along these lines:

https://www.fluentcpp.com/2018/05/08/std-iterator-deprecated/

--- a/src/include/platform/mir/udev/wrapper.h
+++ b/src/include/platform/mir/udev/wrapper.h
@@ -95,9 +95,12 @@ public:
     void match_parent(Device const& parent);
     void match_sysname(std::string const& sysname);
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
     class iterator :
         public std::iterator<std::input_iterator_tag, Device>
     {
+#pragma GCC diagnostic pop
     public:
         iterator& operator++();
         iterator operator++(int);

--- a/src/include/platform/mir/udev/wrapper.h
+++ b/src/include/platform/mir/udev/wrapper.h
@@ -95,13 +95,15 @@ public:
     void match_parent(Device const& parent);
     void match_sysname(std::string const& sysname);
 
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
-    class iterator :
-        public std::iterator<std::input_iterator_tag, Device>
+    class iterator
     {
-#pragma GCC diagnostic pop
     public:
+        using iterator_category = std::input_iterator_tag;
+        using value_type = Device;
+        using difference_type = ptrdiff_t;
+        using pointer = Device*;
+        using reference = Device&;
+
         iterator& operator++();
         iterator operator++(int);
 

--- a/src/platform/graphics/gamma_curves.cpp
+++ b/src/platform/graphics/gamma_curves.cpp
@@ -20,6 +20,7 @@
 
 #include <boost/throw_exception.hpp>
 
+#include <algorithm>
 #include <limits>
 #include <stdexcept>
 

--- a/src/platforms/common/server/kms-utils/drm_mode_resources.h
+++ b/src/platforms/common/server/kms-utils/drm_mode_resources.h
@@ -57,9 +57,12 @@ template<typename DRMUPtr, DRMUPtr(*)(int, uint32_t)>
 class ObjectCollection
 {
 public:
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
     class iterator :
         public std::iterator<std::input_iterator_tag, DRMUPtr>
     {
+#pragma GCC diagnostic pop
     public:
         iterator(iterator const& from);
         iterator& operator=(iterator const& rhs);

--- a/src/platforms/common/server/kms-utils/drm_mode_resources.h
+++ b/src/platforms/common/server/kms-utils/drm_mode_resources.h
@@ -57,13 +57,15 @@ template<typename DRMUPtr, DRMUPtr(*)(int, uint32_t)>
 class ObjectCollection
 {
 public:
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
-    class iterator :
-        public std::iterator<std::input_iterator_tag, DRMUPtr>
+    class iterator
     {
-#pragma GCC diagnostic pop
     public:
+        using iterator_category = std::input_iterator_tag;
+        using value_type = DRMUPtr;
+        using difference_type = ptrdiff_t;
+        using pointer = DRMUPtr*;
+        using reference = DRMUPtr&;
+
         iterator(iterator const& from);
         iterator& operator=(iterator const& rhs);
 

--- a/tests/platform_test_harness/graphics_platform_test_harness.cpp
+++ b/tests/platform_test_harness/graphics_platform_test_harness.cpp
@@ -204,17 +204,6 @@ auto test_display_construction(mir::graphics::DisplayPlatform& platform, Minimal
     }
 }
 
-auto test_display_supports_gl(mg::Display& display) -> bool
-{
-    if (dynamic_cast<mir::renderer::gl::ContextSource*>(&display))
-    {
-        std::cout << "Display supports GL context creation" << std::endl;
-        return true;
-    }
-    std::cout << "Display does not support GL context creation" << std::endl;
-    return false;
-}
-
 void for_each_display_buffer(mg::Display& display, std::function<void(mg::DisplayBuffer&)> functor)
 {
     display.for_each_display_sync_group(
@@ -556,7 +545,6 @@ int main(int argc, char const** argv)
         {
             if (auto display = test_display_construction(*platform, config))
             {
-                success &= test_display_supports_gl(*display);
                 success &= dump_egl_config(*display);
                 success &= test_display_has_at_least_one_enabled_output(*display);
                 success &= test_display_buffers_support_gl(*display);

